vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -0,0 +1,121 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	externalmetricsv1beta1 "k8s.io/metrics/pkg/apis/external_metrics/v1beta1"
+	"k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	"time"
+
+	resourceclient "k8s.io/metrics/pkg/client/clientset/versioned/typed/metrics/v1beta1"
+	"k8s.io/metrics/pkg/client/external_metrics"
+)
+
+type Source interface {
+	List(ctx context.Context, namespace string, model *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error)
+}
+
+type podMetricsSource struct {
+	metricsGetter resourceclient.PodMetricsesGetter
+}
+
+func NewPodMetricsesSource(source resourceclient.PodMetricsesGetter) Source {
+	return podMetricsSource{metricsGetter: source}
+}
+
+func (s podMetricsSource) List(ctx context.Context, namespace string, _ *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	podMetricsInterface := s.metricsGetter.PodMetricses(namespace)
+	return podMetricsInterface.List(ctx, opts)
+}
+
+type externalMetricsClient struct {
+	externalClient external_metrics.ExternalMetricsClient
+	options        ExternalClientOptions
+}
+
+type ExternalClientOptions struct {
+	CpuMetric, MemoryMetric                          string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
+}
+
+func NewExternalClient(c *rest.Config, options ExternalClientOptions) Source {
+	extClient, err := external_metrics.NewForConfig(c)
+	if err != nil {
+		klog.Fatalf("Failed initializing external metrics client: %v", err)
+	}
+	return externalMetricsClient{
+		externalClient: extClient,
+		options:        options,
+	}
+}
+
+func (s externalMetricsClient) containerId(value externalmetricsv1beta1.ExternalMetricValue) *model.ContainerID {
+	podNS, hasPodNS := value.MetricLabels[s.options.PodNamespaceLabel]
+	podName, hasPodName := value.MetricLabels[s.options.PodNameLabel]
+	ctrName, hasCtrName := value.MetricLabels[s.options.CtrNameLabel]
+	if hasPodNS && hasPodName && hasCtrName {
+		return &model.ContainerID{
+			PodID:         model.PodID{Namespace: podNS, PodName: podName},
+			ContainerName: ctrName,
+		}
+	} else {
+		return nil
+	}
+}
+
+func (s externalMetricsClient) List(ctx context.Context, namespace string, state *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	result := v1beta1.PodMetricsList{}
+	// Get all VPAs in the namespace
+	// - We already do this in the cluster state feeder!  It's in its clusterState member.
+	//   We just have to feed it into here somehow.
+	// - use the 'PodSelector' there as the input to the external api.
+	// Send out the queries.
+	nsClient := s.externalClient.NamespacedMetrics(namespace)
+	for _, vpa := range state.Vpas {
+		workloadValues := make(map[model.PodID]map[string]k8sapiv1.ResourceList)
+		cpuMetrics, err := nsClient.List(s.options.CpuMetric, vpa.PodSelector)
```

Comment:
The code for memory is a copy of the code for CPU, create a method which will get passed the metric type and name, and call it once for cpu, once for memory

Reply:
I don't think this has been addressed no ? I think we can create a helper method that fetches, does the err check / empty check, calls `s.addMetrics` and then just iterate over the metric types no ? (which will be nice when we will want to implement it for network)